### PR TITLE
Add State Views and Config View specialization

### PR DESCRIPTION
**Pull request by peterschwarz**
**Add StateView and its Factory**

StateViewFactory provides read-only views into state stored in a merkle tree database. These views can provide a window into the state which can be used throughout a block of processing without concern of the state being changed underneath the process.

**Add ConfigView and its Factory**

ConfigView is a specialized StateView which translates keys into state addresses and transforms state entries into their respective values.  This can be used by validator components to read on-chain settings at viewed from the lens of a specific point in the chain.

### CODE REVIEW

File: validator/sawtooth_validator/state/state_view.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        """
+        Returns:
+            list of str: the list of addresses available in this view
+        """
+        return self._tree.addresses()
+
+    def leaves(self, prefix):
+        """
+        Args:
+            prefix (str): an address prefix under which to look fo leaves
```
Comment (jsmitchell) [inline validator/sawtooth_validator/state/state_view.py]: s/fo/for

**Comment by cintel** [comment]
Probably a nit but shouldn't all of these files have a new line at the end?

**Comment by peterschwarz** [comment]
new-lines would fail our linting rules. But, since #79 is not merged in yet, it did not catch some other linting issues that are there.  Will fix.

nit away!

**Comment by peterschwarz** [comment]
Hold off on this - need to correct some tests

**Pull request closed without merging**